gax/src/main/java/com/google/api/gax/rpc/ServerStreamingAttemptCallable.java

```diff
@@ -0,0 +1,254 @@
+              );
+            }
+          }
+
+          @Override
+          public void onCompleteImpl() {
+            attemptFuture.setPrivately(null);
+          }
+        },
+        context);
+
+    externalFuture.setAttemptFuture(attemptFuture);
+
+    return null;
+  }
+
+  /**
+   * Called by the inner {@link ServerStreamingCallable} when the call is about to start. This will
+   * transfer unfinished state from the previous attempt.
+   *
+   * @see ResponseObserver#onStart(StreamController)
+   */
+  private void onAttemptStart(StreamController controller) {
+    if (!autoFlowControl) {
+      controller.disableAutoInboundFlowControl();
+    }
+
+    Throwable localCancellationCause;
+    int numToRequest = 0;
+
+    synchronized (lock) {
+      innerController = controller;
+
+      localCancellationCause = this.cancellationCause;
+
+      if (!autoFlowControl) {
+        numToRequest = pendingRequests;
+      }
+    }
+
+    if (localCancellationCause != null) {
+      controller.cancel();
+    } else if (numToRequest > 0) {
+      controller.request(numToRequest);
+    }
+  }
+
+  /**
+   * Called when the outer {@link ResponseObserver} is ready for more data.
+   *
+   * @see StreamController#request(int)
+   */
+  private void onRequest(int count) {
+    Preconditions.checkState(!autoFlowControl, "Automatic flow control is enabled");
+    Preconditions.checkArgument(count > 0, "Count must be > 0");
+
+    final StreamController localInnerController;
+
+    synchronized (lock) {
+      int maxInc = Integer.MAX_VALUE - pendingRequests;
+      count = Math.min(maxInc, count);
+
+      pendingRequests += count;
+      localInnerController = this.innerController;
+    }
+
+    // Note: there is a race condition here where the count might go to the previous attempt's
+    // StreamController after it failed. But it doesn't matter, because the controller will just
+    // ignore it and the current controller will pick it up onStart.
+    if (localInnerController != null) {
+      localInnerController.request(count);
+    }
+  }
+
+  /**
+   * Called when the outer {@link ResponseObserver} wants to prematurely cancel the stream.
+   *
+   * @see StreamController#cancel()
+   */
+  private void onCancel() {
+    StreamController localInnerController;
+
+    synchronized (lock) {
+      if (cancellationCause != null) {
+        return;
+      }
+      cancellationCause = new WrappedCancellationException("User cancelled stream");
+      localInnerController = innerController;
+    }
+
+    if (localInnerController != null) {
+      localInnerController.cancel();
+    }
+  }
+
+  public static class WrappedCancellationException extends RuntimeException {
+    public WrappedCancellationException(String message) {
+      super(new CancellationException(message));
+    }
+  }
+
+  public static class WrappedApiException extends RuntimeException {
```

Comment:
"wrapped" doesn't really have any meaning... especially because there isn't really even an object there that is wrapped. Maybe name it something like "StreamingAttemptException" or something along those lines.

Reply:
will do